dev/TeachingTip/TeachingTip.cpp

```diff
@@ -710,28 +720,46 @@ void TeachingTip::OnIsOpenChanged()
             this->ApplyTemplate();
         }
 
-        if (!m_popup)
+        if (!m_popup || m_createNewPopupOnOpen)
         {
             auto popup = winrt::Popup();
             m_popupOpenedRevoker = popup.Opened(winrt::auto_revoke, { this, &TeachingTip::OnPopupOpened });
             m_popupClosedRevoker = popup.Closed(winrt::auto_revoke, { this, &TeachingTip::OnPopupClosed });
+            popup.ShouldConstrainToRootBounds(ShouldConstrainToRootBounds());
             m_popup.set(popup);
+            m_createNewPopupOnOpen = false;
         }
 
-        auto&& popup = m_popup.get();
-        if (!popup.IsOpen())
+        // If the tip is not going to open because it does not fit we need to make sure that
+        // the open, closing, closed life cycle still fires so that we don't cause apps to leak
+        // that depend on this sequence.
+        DetermineEffectivePlacement();
+        if (m_tipDoesNotFit)
         {
```

Comment:
When I see member fields tracking transient calculations like this, it's a red flag for me (violation of [ES.5 Keep scopes small](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es5-keep-scopes-small)). Can this be a value returned by DetermineEffectivePlacement?

Reply:
And even if this value is needed, I still think that DetermineEffectivePlacement should not have side effects of setting fields. I see that PositionPopup needs this field but I don't understand the flow of code enough to recommend a way to avoid the field. Who calls PositionPopup?

---
In reply to: [268778409](https://github.com/Microsoft/microsoft-ui-xaml/pull/480#discussion_r268778409) [](ancestors = 268778409)